feat(AIP-3) add versioning

toumorokoshi opened the pull request:
Adding an AIP to introduce versioning,
to help API authors point at older AIP guidance
as it evolves.

fixes #1054

toumorokoshi committed:
feat(AIP-3) add versioning

Adding an AIP to introduce versioning,
to help API authors point at older AIP guidance
as it evolves.

fixes #1054

shwoodard commented on aip/general/0003.md:
```diff
@@ -0,0 +1,34 @@
+
+# AIP Versioning
+
+This AIP defines the versioning scheme of the AIPs.
```
I don't see many or any AIPs that start with "This AIP...". Please rephrase.

toumorokoshi replied:
https://google.aip.dev/9 is an example.

Is this guidance stated anywhere? What do you see as the disadvantage of using this form?

I'm okay with changing the text, I just think we should be clear on requirements vs trying to infer conventions.

shwoodard commented on aip/general/0003.md:
```diff
@@ -0,0 +1,34 @@
+
+- The AIPs **must** be versioned by date, using the [ISO-8601 format of `YYYY-MM-DD`](https://www.w3.org/TR/NOTE-datetime), that corresponds to the date the version was added.
+- AIPs versions **must** be available as a tag on the source control mechanism used to store the AIPs, of the format `v{version}`. Example: `v2023-03-28`.
+- The AIPs **should** have a new version every 3 months, or when a significant change to an AIP occurs.
```
Line length. Please use some form of auto-formatting.

shwoodard commented on aip/general/0003.md:
```diff
@@ -0,0 +1,34 @@
+
+## Guidance
+
+- The AIPs **must** be versioned by date, using the [ISO-8601 format of `YYYY-MM-DD`](https://www.w3.org/TR/NOTE-datetime), that corresponds to the date the version was added.
```
I thought the proposal was `yyyy-MM`--not to include day of month.

toumorokoshi replied:
please see the issue comments: @jskeet suggested ISO-8601 `YYYY-MM-DD`.

toumorokoshi replied:
Note this is also stated in the appendix.

shwoodard commented on aip/general/0003.md:
```diff
@@ -0,0 +1,34 @@
+## Guidance
+
+- The AIPs **must** be versioned by date, using the [ISO-8601 format of `YYYY-MM-DD`](https://www.w3.org/TR/NOTE-datetime), that corresponds to the date the version was added.
+- AIPs versions **must** be available as a tag on the source control mechanism used to store the AIPs, of the format `v{version}`. Example: `v2023-03-28`.
```
Nit: s/mechanism/provider/

toumorokoshi replied:
would you mind clarifying why you think that's clearer?

shwoodard replied:
Source control is a mechanism, so it is confusing. `provider` or, better yet, `system` differentiates one src control technology from another.

toumorokoshi replied:
I was thinking "technology" was good one too. But I agree system is clearer than mechanism, will fix!

shwoodard commented on aip/general/0003.md:
```diff
@@ -0,0 +1,34 @@
+
+- The AIPs **must** be versioned by date, using the [ISO-8601 format of `YYYY-MM-DD`](https://www.w3.org/TR/NOTE-datetime), that corresponds to the date the version was added.
+- AIPs versions **must** be available as a tag on the source control mechanism used to store the AIPs, of the format `v{version}`. Example: `v2023-03-28`.
+- The AIPs **should** have a new version every 3 months, or when a significant change to an AIP occurs.
```
Why not **must**? Does this mean that the version of the AIP should be changed every three months regardless of it is altered?

toumorokoshi replied:
This is explained in the issue, @jskeet was saying there may be situations where we don't need a new version.

toumorokoshi replied:
But this should be added into the Appendix since it's not clear in the document. I'll make sure to add that.

toumorokoshi replied:
For now I removed the *every three months* part due to some feedback from @rofrankel in the issue, and focused more on versioning based on "significant changes" without a precise definition.

I'll make it a must now.

shwoodard commented on aip/general/0003.md:
```diff
@@ -0,0 +1,34 @@
+
+## Guidance
+
+- The AIPs **must** be versioned by date, using the [ISO-8601 format of `YYYY-MM-DD`](https://www.w3.org/TR/NOTE-datetime), that corresponds to the date the version was added.
```
How? Is there text in the metadata table at the top of the AIP that has the version? Update the template in this PR as well?

toumorokoshi replied:
it's a tag in the version control, please see the following line.

I don't view this as being a version of a specific AIP, as the guidance of a single AIP is generally dependent on other AIPs (e.g. around standard methods). 

WDYT?

shwoodard commented on aip/general/0003.md:
```diff
@@ -0,0 +1,34 @@
+The rational for date-based versioning is:
+
+- It allows for a simple heuristic to correlate the version of AIPs an API followed (look at the date of the API change, then look at the AIP version published right before that date).
+- The daily granularity enables multiple changes to be published in a single month.
```
Use standard line length.

shwoodard commented on aip/general/0003.md:
```diff
@@ -0,0 +1,34 @@
+As the AIPs continue to evolve, it is beneficial for
+the AIPs to have a way to denote older versions of the
+specification to serve as a reference point for APIs
+to explain the rationale of their design choices.
```
Combine the two paragraphs to read,

> As AIPs evolve, versioning is beneficial. Versions serve as reference points that demonstrate the rationale for design choices. They are crucial since sometimes the guidance within an AIP is reversed or includes updates to design patterns that may not look similar to the original design.

toumorokoshi replied:
Can you pinpoint the parts you think are valuable relative to the original paragraph?

I'm not sure I see how the content in your paragraph differs from the content in the original one. So trying to figure out what I'm missing.

shwoodard replied:
I'm confused by your response. You talk about a singular "paragraph". I am suggesting combing this paragraph with the one below, thus eliminating the second paragraph.

In fact, in my edit, s/are crucial since/are crucial as/.

Less is more. No meaning is lost by combining the paragraphs and the readability is improved.

toumorokoshi replied:
ack, I agree the paragraphs are more verbose than they need to be. I'll push an update that shrinks to a single paragraph while keeping a point I think is valuable (clients referencing older versions).

toumorokoshi replied:
PTAL.

shwoodard commented on aip/general/0003.md:
```diff
@@ -0,0 +1,34 @@
+
+The rational for date-based versioning is:
+
+- It allows for a simple heuristic to correlate the version of AIPs an API followed (look at the date of the API change, then look at the AIP version published right before that date).
```
> It provides a simple means to order changes to AIPs. Since the changes are listed in the change log at the bottom of each AIP, simply compare the dates their with version date to see the previous version.

toumorokoshi replied:
I think your rewrite is saying something a bit different from what I'm saying: in this proposal, the individual AIPs do not have a version. The total set of AIPs do.

So my statement is that one could pull up the version of this repository with the particular stamp to see the full AIP guidance at that time. 

Viewing the changelog is also valuable, so I definitely will add that note in the rationale section!

shwoodard replied:
Ok, edit as you see fit, however,

1. This is a runon sentence: sentences need to follow grammatical rules with or without parentheticals.
2. "to correlate the version of AIPs an API follows" is cumbersome and difficult to understand. Please fix.

toumorokoshi replied:
ack. agreed this longer than it needs to be. I'm opting to be more terse and hoping people can read between the lines a little bit. PTAL.

toumorokoshi committed:
addressing feedback

- fixing line lengths
- discussing alternatives like fine-grained AIP versioning
- adding guidance on changelogs.

toumorokoshi committed:
addressing feedback

- eliminate run-on sentences.
- make line length consistent.
- use "source control system" instead of "mechanism".

toumorokoshi committed:
Merge branch 'master' into toum/add-versioning

shwoodard commented on aip/general/0003.md:
```diff
@@ -0,0 +1,45 @@
+### Rationale for date-based versioning
+
+With date-based versioning, a client can easily find the AIP guidance at the
+time an API was authored.
```
Add one new line below here.

shwoodard commented on aip/general/0003.md:
```diff
@@ -0,0 +1,45 @@
+time an API was authored.
+### Why not individually versioned AIPs
+
+An alternative to a universal version is having specific versions attached to
```
s/having/to have/

passive voice -> active voice

shwoodard commented on aip/general/0003.md:
```diff
@@ -0,0 +1,45 @@
+An alternative to a universal version is having specific versions attached to
+each AIP.
+
+This approach was not opted due to the fact that the AIPs often cross-link and
```
s/opted/chosen/g

I don't think this is a complete sentence as is.

toumorokoshi replied:
ack. I rewrote the section to be a little more terse and try to convey simpler points in each sentence. PTAL.

shwoodard commented on aip/general/0003.md:
```diff
@@ -0,0 +1,45 @@
+
+This approach was not opted due to the fact that the AIPs often cross-link and
+depend on the rules in others (e.g. many rely on the resource-oriented AIPs).
+Therefore linking to a specific version of a single AIP would not be a complete
```
Add a comma after "Therefore".

toumorokoshi committed:
addressing feedback:

- renaming appendix to rationale, since that's
  the new name of the section.
- wording fixes
- trying to have terser rationale.

toumorokoshi committed:
Merge branch 'master' into toum/add-versioning

toumorokoshi committed:
addressing feedback

other -> one another

bgrant0607 reviewed (approved):
LGTM
